validphys2/src/validphys/utils.py

```diff
@@ -5,14 +5,52 @@
 @author: Zahari Kassabov
 """
 import contextlib
+import functools
 import pathlib
 import shutil
 import tempfile
+from typing import Any, Sequence, Mapping
 
 import numpy as np
 from validobj import ValidationError, parse_input
 
 
+# Since typing.Hashable doesn't check recursively you actually
+# have to try hashing it.
+def is_hashable(obj):
+    try:
+        hash(obj)
+        return True
+    except:
+        return False
+
+
+def immute(obj: Any):
+    # So that we don't infinitely recurse since frozenset and tuples
+    # are Sequences.
+    if is_hashable(obj):
+        return obj
+    elif isinstance(obj, Mapping):
+        return frozenset([(immute(k), immute(v)) for k, v in obj.items()])
```

Comment:
The dict keys are already hashable.

```suggestion
        return frozenset([k, immute(v)) for k, v in obj.items()])
```